server/services/ai.ts

```diff
@@ -1,33 +1,185 @@
+import { AiState, KnownAttack } from "../types/Ai";
+import CarrierService from "./carrier";
+import CombatService from "./combat";
+import DistanceService from "./distance";
+import PlayerService from "./player";
+import ShipTransferService from "./shipTransfer";
+import StarService from "./star";
 import StarUpgradeService from "./starUpgrade";
+import TechnologyService from "./technology";
+import WaypointService from "./waypoint";
+import { Star } from "../types/Star";
+import { Carrier } from "../types/Carrier";
+import {getOrInsert, reverseSort, notNull, maxBy} from "../utils";
+import {CarrierWaypoint, CarrierWaypointActionType} from "../types/CarrierWaypoint";
+import ReputationService from "./reputation";
+import DiplomacyService from "./diplomacy";
+import PlayerStatisticsService from "./playerStatistics";
 
 const FIRST_TICK_BULK_UPGRADE_SCI_PERCENTAGE = 20;
 const FIRST_TICK_BULK_UPGRADE_IND_PERCENTAGE = 30;
 const LAST_TICK_BULK_UPGRADE_ECO_PERCENTAGE = 100;
 
-export default class AIService {
+const Heap = require('qheap');
+const mongoose = require("mongoose");
+
+enum AiAction {
+    DefendStar,
+    ClaimStar,
+    ReinforceStar,
+    InvadeStar
+}
+
+interface DefendStarOrder {
+    type: AiAction.DefendStar,
+    score: number,
+    star: string,
+    ticksUntil: number,
+    incomingCarriers: Carrier[]
+}
+
+interface ClaimStarOrder {
+    type: AiAction.ClaimStar,
+    star: string,
+    score: number
+}
+
+interface ReinforceStarOrder {
+    type: AiAction.ReinforceStar,
+    score: number,
+    star: string,
+    source: string
+}
+
+interface InvadeStarOrder {
+    type: AiAction.InvadeStar,
+    star: string,
+    score: number
+}
+
+interface TracePoint {
+    starId: string,
+    action?: CarrierWaypointActionType
+}
+
+type Order = DefendStarOrder | ClaimStarOrder | ReinforceStarOrder | InvadeStarOrder;
+
+type StarGraph = Map<string, Set<string>>;
+
+interface Context {
+    aiState: AiState;
+    playerStars: Star[];
+    playerCarriers: Carrier[];
+    starsById: Map<string, Star>;
+    allReachableFromPlayerStars: StarGraph;
+    freelyReachableFromPlayerStars: StarGraph;
+    reachablePlayerStars: StarGraph;
+    freelyReachableStars: StarGraph;
+    allCanReachPlayerStars: StarGraph;
+    starsInGlobalRange: StarGraph;
+    borderStars: Set<string>;
+    carriersOrbiting: Map<string, Carrier[]>;
+    carriersById: Map<string, Carrier>;
+    attacksByStarId: Map<string, Map<number, Carrier[]>>;
+    attackedStarIds: Set<string>;
+    playerEconomy: number;
+    playerIndustry: number;
+    playerScience: number;
+}
+
+interface Assignment {
+    carriers: Carrier[];
+    star: Star;
+    totalShips: number;
+}
+
+interface FoundAssignment {
+    assignment: Assignment;
+    trace: TracePoint[];
+}
+
+const EMPTY_STAR_SCORE_MULTIPLIER = 1;
+const ENEMY_STAR_SCORE_MULTIPLIER = 5;
+
+const REINFORCEMENT_MIN_CYCLES = 1.5;
+const REINFORCEMENT_MIN_FACTOR = 1.4;
 
+const INVASION_ATTACK_FACTOR = 1.5;
```

Comment:
We should add these to game constants *at some point*, not necessary for MVP.

Reply:
Agreed. They also might get some more fine tuning anyways, or rather, some of my further ideas would largely remove the need for the constants used at the moment.